knitr/math-guide/math_guide.Rmd

```diff
@@ -0,0 +1,643 @@
+the arena.
+
+`to_arena` is a function that takes one argument (usually an Eigen vector
+or matrix) and returns a lightweight object pointing to memory on the
+arena.
+
+`arena_t<T>` defines, for a given type `T`, the type of a lightweight
+object pointing to memory in the arena that can store a `T`.
+
+Copying a variable `x` to the arena can be done with either `to_arena` or
+`arena_t<T>`:
+
+```cpp
+template <typename T>
+auto myfunc(const T& x) {
+  auto arena_x = to_arena(x);
+}
+```
+
+```cpp
+template <typename T>
+auto myfunc(const T& x) {
+  arena_t<T> arena_x = x;
+}
+```
+
+### Array of Structures vs. Structure of Array types (AOS vs. SOA, matvar vs. varmat)
+
+There are two types of reverse mode autodiff in Math. They are colloquially
+referred to as matvar and varmat, though it is more accurate to
+refer to them as Array of Structure (AOS) and Structure of Array (SOA)
+autodiff types.
+
+The first autodiff type in Stan was `var`. A `var` is the autodiff version
+of a `double`. Constructing matrix autodiff types with `var` is as simple
+as building an Eigen matrix and filling it with vars, so
+`Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>`. This is an AOS
+autodiff style.
+
+The SOA autodiff style defines vector and matrices themselves as autodiff
+variables. The SOA matrix is `var_value<Eigen::MatrixXd>`. Defining an
+entire matrix as an autodiff type makes it possible to do a number of
+optimizations that are not possible when the autodiff types are just
+variables stored inside another structure.
+
+### Reverse mode `require` metaprograms
+
+Require metaprograms are useful for avoiding ambiguities between fully
+templated prim implementations and reverse mode specializations.
+
+The simplest way to do this is add a `require_any_st_var<...>` to
+the reverse mode implementation and a `require_all_not_st_var<...>` to
+the prim implementation.
+
+### Values
+
+`value_of(x)` returns the values of the variable `x`. If `x` is not an
+autodiff type, it simply forwards `x` along. The values of `x` have
+the same shape as `x`. For reverse mode autodiff, values are `double`
+variables. For higher order autodiff (`fvar<var>`, etc.) this is not
+always the case. See `value_of_rec` for a different behavior.
+
+### Values and adjoint extensions to Eigen
+
+The AOS matrix and vector autodiff types come with an extra `.val()`,
+`.val_op()`, `.adj()`, and `.adj_op()` member functions.
+
+`.val()` and `.val_op()` return expressions that evaluate to the values
+of the autodiff matrix.
+
+`.adj()` and `.adj_op()` return expressions that evaluate to the adjoints
+of the autodiff matrix.
+
+The non-`_op` versions of the functions should be preferred. If the
+code does not compile (usually an error about const-ness) try the `_op`
+versions. This can happen when multiplying the values/adjoints of an AOS
+autodiff type against other matrices.
+
+It is not safe to use `.noalias()` with the AOS value/adjoint expressions.
```

Comment:
Why not?

Reply:
I'm not sure, but I believe it exploded when I did it before. When @SteveBronder did this he added some if statements so the noalias code only ran for SOA. I forget where it was though (I thought multiply but when I looked it wasn't there).